runners/core-java/src/main/java/org/apache/beam/runners/core/triggers/TriggerStateMachines.java

```diff
@@ -0,0 +1,215 @@
+      return AfterWatermarkStateMachine.pastEndOfWindow();
+    }
+
+    private OnceTriggerStateMachine evaluateSpecific(NeverTrigger v) {
+      return NeverStateMachine.ever();
+    }
+
+    private OnceTriggerStateMachine evaluateSpecific(AfterSynchronizedProcessingTime v) {
+      return new AfterSynchronizedProcessingTimeStateMachine();
+    }
+
+    private OnceTriggerStateMachine evaluateSpecific(AfterFirst v) {
+      List<OnceTriggerStateMachine> subStateMachines =
+          Lists.newArrayListWithCapacity(v.subTriggers().size());
+      for (Trigger subtrigger : v.subTriggers()) {
+        subStateMachines.add(stateMachineForOnceTrigger((OnceTrigger) subtrigger));
+      }
+      return AfterFirstStateMachine.of(subStateMachines);
+    }
+
+    private OnceTriggerStateMachine evaluateSpecific(AfterAll v) {
+      List<OnceTriggerStateMachine> subStateMachines =
+          Lists.newArrayListWithCapacity(v.subTriggers().size());
+      for (Trigger subtrigger : v.subTriggers()) {
+        subStateMachines.add(stateMachineForOnceTrigger((OnceTrigger) subtrigger));
+      }
+      return AfterAllStateMachine.of(subStateMachines);
+    }
+
+    private OnceTriggerStateMachine evaluateSpecific(AfterPane v) {
+      return AfterPaneStateMachine.elementCountAtLeast(v.getElementCount());
+    }
+
+    private TriggerStateMachine evaluateSpecific(AfterWatermark.AfterWatermarkEarlyAndLate v) {
+      AfterWatermarkStateMachine.AfterWatermarkEarlyAndLate machine =
+          AfterWatermarkStateMachine.pastEndOfWindow()
+              .withEarlyFirings(stateMachineForOnceTrigger(v.getEarlyTrigger()));
+
+      if (v.getLateTrigger() != null) {
+        machine = machine.withLateFirings(stateMachineForOnceTrigger(v.getLateTrigger()));
+      }
+      return machine;
+    }
+
+    private TriggerStateMachine evaluateSpecific(AfterEach v) {
+      List<TriggerStateMachine> subStateMachines =
+          Lists.newArrayListWithCapacity(v.subTriggers().size());
+
+      for (Trigger subtrigger : v.subTriggers()) {
+        subStateMachines.add(stateMachineForTrigger(subtrigger));
+      }
+
+      return AfterEachStateMachine.inOrder(subStateMachines);
+    }
+
+    private TriggerStateMachine evaluateSpecific(Repeatedly v) {
+      return RepeatedlyStateMachine.forever(stateMachineForTrigger(v.getRepeatedTrigger()));
+    }
+
+    private TriggerStateMachine evaluateSpecific(OrFinallyTrigger v) {
+      return new OrFinallyStateMachine(
+          stateMachineForTrigger(v.getMainTrigger()),
+          stateMachineForOnceTrigger(v.getUntilTrigger()));
+    }
+
+    private OnceTriggerStateMachine evaluateSpecific(AfterProcessingTime v) {
+      return evaluateSpecific((AfterDelayFromFirstElement) v);
+    }
+
+    private OnceTriggerStateMachine evaluateSpecific(final AfterDelayFromFirstElement v) {
+      return new AfterDelayFromFirstElementStateMachineAdapter(v);
+    }
+
+    private TriggerStateMachine evaluateUnknown(Trigger trigger) {
```

Comment:
I think it might be dead now. Previously it caught missing triggers and crashed. In an object algebra world, triggers outside the closed set of known triggers used this. Anyhow, I'll get rid of it.